tsl/src/continuous_aggs/union_view.c

```diff
@@ -0,0 +1,316 @@
+							InvalidOid,
+							InvalidOid,
+							COERCE_EXPLICIT_CALL);
+
+	return build_conversion_call(type, boundary);
+}
+
+static Node *
+build_union_query_quals(int32 ht_id, Oid partcoltype, Oid opno, int varno, AttrNumber attno)
+{
+	Var *var = makeVar(varno, attno, partcoltype, -1, InvalidOid, InvalidOid);
+	FuncExpr *boundary = build_boundary_call(ht_id, partcoltype);
+
+	CoalesceExpr *coalesce = makeNode(CoalesceExpr);
+	coalesce->coalescetype = partcoltype;
+	coalesce->coalescecollid = InvalidOid;
+	coalesce->args = list_make2(boundary, make_lower_bound(partcoltype));
+
+	Expr *op = make_opclause(opno,
+							 BOOLOID,
+							 false,
+							 (Expr *) var,
+							 (Expr *) coalesce,
+							 InvalidOid,
+							 InvalidOid);
+
+	return (Node *) op;
+}
+
+static RangeTblEntry *
+make_subquery_rte(Query *subquery, const char *aliasname)
+{
+	RangeTblEntry *rte = makeNode(RangeTblEntry);
+	ListCell *lc;
+
+	rte->rtekind = RTE_SUBQUERY;
+	rte->relid = InvalidOid;
+	rte->subquery = subquery;
+	rte->alias = makeAlias(aliasname, NIL);
+	rte->eref = copyObject(rte->alias);
+
+	foreach (lc, subquery->targetList)
+	{
+		TargetEntry *tle = lfirst_node(TargetEntry, lc);
+		if (!tle->resjunk)
+			rte->eref->colnames = lappend(rte->eref->colnames, makeString(pstrdup(tle->resname)));
+	}
+
+	/*
+	 * Set flags and access permissions.
+	 *
+	 * Subqueries are never checked for access rights.
+	 */
+	rte->lateral = false;
+	rte->inh = false; /* never true for subqueries */
+	rte->inFromCl = true;
+
+	rte->requiredPerms = 0;
+	rte->checkAsUser = InvalidOid;
+	rte->selectedCols = NULL;
+	rte->insertedCols = NULL;
+	rte->updatedCols = NULL;
+
+	return rte;
+}
+
+Query *
+cagg_build_union_query(CAggTimebucketInfo *tbinfo, MatTableColumnInfo *mattblinfo, Query *q1,
+					   Query *q2)
+{
+	ListCell *lc1, *lc2;
+	List *col_types = NIL;
+	List *col_typmods = NIL;
+	List *col_collations = NIL;
+	List *tlist = NIL;
+	int varno;
+	AttrNumber attno;
+
+	Assert(list_length(q1->targetList) == list_length(q2->targetList));
+
+	q1 = copyObject(q1);
+	q2 = copyObject(q2);
+
+	TypeCacheEntry *tce = lookup_type_cache(tbinfo->htpartcoltype, TYPECACHE_LT_OPR);
+
+	varno = list_length(q1->rtable);
+	attno = mattblinfo->matpartcolno + 1;
+	q1->jointree->quals =
+		build_union_query_quals(tbinfo->htid, tbinfo->htpartcoltype, tce->lt_opr, varno, attno);
+	attno =
+		get_attnum(tbinfo->htoid, get_attname_compat(tbinfo->htoid, tbinfo->htpartcolno, false));
+	varno = list_length(q2->rtable);
+	q2->jointree->quals = build_union_query_quals(tbinfo->htid,
+												  tbinfo->htpartcoltype,
+												  get_negator(tce->lt_opr),
+												  varno,
+												  attno);
+
+	Query *query = NULL;
+	SetOperationStmt *setop = makeNode(SetOperationStmt);
+	RangeTblEntry *rte_q1 = make_subquery_rte(q1, "*SELECT* 1");
+	RangeTblEntry *rte_q2 = make_subquery_rte(q2, "*SELECT* 2");
+	RangeTblRef *ref_q1 = makeNode(RangeTblRef);
+	RangeTblRef *ref_q2 = makeNode(RangeTblRef);
```

Comment:
check that length of rte->reref->colnames is the same for rte_q1 and rte_q2 and error out otherwise (to catch unexpected issues).

Reply:
at the start of the function is an assert that both target lists have the same length